src/pudl/transform/vceregen.py

```diff
@@ -0,0 +1,206 @@
+    state_names = cleanstrings_snake(
+        ps_usa_df, ["subdivision_name"]
+    ).subdivision_name.tolist()
+    state_pattern = "|".join(state_names)
+    # Transform lat long fips table by making the county_state_names lowercase
+    # to match the values in the capacity factor tables. Fix FIPS codes with
+    # no leading zeros, and add a unique country and state field.
+    lat_long_fips = (
+        raw_vcegen__lat_lon_fips.pipe(simplify_columns)
+        .assign(county_state_names=lambda x: x.county_state_names.str.lower())
+        .assign(fips=lambda x: zero_pad_numeric_string(x.fips, 5))
+        .assign(
+            county=lambda x: x.county_state_names.str.extract(
+                rf"(.+)_({state_pattern})"
+            )[0]
+        )
+        .assign(
+            state=lambda x: x.county_state_names.str.extract(
+                rf"(.+)_({state_pattern})"
+            )[1]
+        )
+        .rename(
+            columns={
+                "lat_county": "latitude",
+                "long_county": "longitude",
+                "fips": "county_id_fips",
+            }
+        )
+    )
+
+    return lat_long_fips
+
+
+def _add_time_cols(df: pd.DataFrame) -> pd.DataFrame:
+    """Add datetime and year_hour columns.
+
+    This function adds a datetime column and a year_hour column.
+    The datetime column is important for merging the data with other
+    data, and the year_hour 0-8769 column is important for modeling
+    purposes. The report_year column is also helpful for filtering,
+    so we keep all three!
+
+    For leap years (2020), December 31st is excluded.
+
+    TO-DO: decide whether to have the year_hour start at 0 or 1.
+    and update the date_range accordingly.
+    """
+    logger.info("Adding datetime columns")
+    # This data is compiled for modeling purposes and skips the last
+    # day of a leap year. If we want to add a datetime column, we need
+    # to make sure that we skip the 31st of December, 2020.
+    # the [:-1] is because datetime will include both the 0th hour of the
+    # start date and the 0th hour of the end date. We only want one!
+    # Though we still need to decide which. This at least makes the math
+    # right.
+    to_leap_year_dates = pd.date_range(start="2019-01-01", end="2020-12-31", freq="h")[
+        :-1
+    ]
+    post_leap_year_dates = pd.date_range(
+        start="2021-01-01", end="2024-01-01", freq="h"
+    )[:-1]
+    # Add date sequence to pre and post leap year dates
+    df.loc[:, "report_year"] = df.report_year.astype("Int32")
+    df.loc[df["report_year"] < 2021, "hour"] = to_leap_year_dates
+    df.loc[df["report_year"] > 2020, "hour"] = post_leap_year_dates
+    # Add a year_date column for modeling purposes.
+    # Right now each year should go from 0-8759. We
+    # can decide to make that 1-8760 if we want to!
+    df.loc[:, "year_hour"] = df.hour.dt.hour + (df.hour.dt.dayofyear - 1) * 24
```

Comment:
do we want this to reset every year (i.e.: 0-8759) or do we want it to go strait through the years (i.e.: 0-43800)

Reply:
Given that the files already come with an hour index that starts with 1, I think we should start with 1.

The hour-of-year should reset every year.

Reply:
Given that the data comes with the hour-of-year we should make sure that we aren't altering it and that the datetime we've created is consistent with it. Either by constructing the datetime from it, or by checking that the provided hour-of-year matches what we would derive from our datetime.

When I suggested creating a datetime column and deriving the hour-of-year from it I thought that we had no indication of the time other than the filename (for the year) and the row number (for the hour).

Reply:
`hour_of_year` feels like it's more consistent with the `day_of_year` and `day_of_week` naming that pandas uses for counting subdivisions.